From 84d87cfdd9f2be58fb617db84501a319ccd0de38 Mon Sep 17 00:00:00 2001
From: Alaric Senat <dev.asenat@posteo.net>
Date: Mon, 27 May 2024 14:15:42 +0200
Subject: [PATCH 52/54] upnp: implement root MRL forging with std::string

Switching to std::string simplifies the next patch changing the URL
scheme.

(cherry picked from commit bc1d58c065d1a359863f3bc4216351b54c60e51d)
---
 modules/services_discovery/upnp.cpp | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/modules/services_discovery/upnp.cpp b/modules/services_discovery/upnp.cpp
index 323f6e5bcd..fc5d93ca2b 100644
--- a/modules/services_discovery/upnp.cpp
+++ b/modules/services_discovery/upnp.cpp
@@ -350,17 +350,17 @@ bool MediaServerList::addServer( MediaServerDesc* desc )
             free( psz_playlist_option );
         }
     } else {
-        char* psz_mrl;
-        // We might already have some options specified in the location.
-        char opt_delim = desc->location.find( '?' ) == std::string::npos ? '?' : '&';
-        if( asprintf( &psz_mrl, "upnp://%s%cObjectID=0", desc->location.c_str(), opt_delim ) < 0 )
-            return false;
+        std::string mrl = std::string("upnp://") + desc->location;
+
+        // Forge a root object ID in the MRL, this is used in the dir access.
+        if ( desc->location.find( '?' ) == std::string::npos )
+            mrl += "?ObjectID=0";
+        else 
+            mrl += "&ObjectID=0";
 
-        p_input_item = input_item_NewDirectory( psz_mrl,
+        p_input_item = input_item_NewDirectory( mrl.c_str(),
                                                 desc->friendlyName.c_str(),
                                                 ITEM_NET );
-        free( psz_mrl );
-
         if ( !p_input_item )
             return false;
 
-- 
2.39.3 (Apple Git-146)

